src/main/java/se/liu/ida/hefquin/engine/queryproc/impl/loptimizer/ApplyVocabularyMappings.java

```diff
@@ -67,14 +47,44 @@ public static LogicalPlan rewriteToUseLocalVocabulary( final LogicalPlan inputPl
 	public LogicalPlan apply( final LogicalPlan inputPlan ) {
 		if (inputPlan.getRootOperator() instanceof LogicalOpRequest) {
 			final LogicalOpRequest<?,?> requestOp = (LogicalOpRequest<?,?>) inputPlan.getRootOperator();
-			if(requestOp.getFederationMember().getVocabularyMapping() != null) { // If fm has a vocabulary mapping vm
-				final LogicalOpLocalToGlobal l2g = new LogicalOpLocalToGlobal(requestOp.getFederationMember().getVocabularyMapping());
-				final LogicalPlan rw = rewriteToUseLocalVocabulary(inputPlan);
-				return new LogicalPlanWithUnaryRootImpl(l2g,rw);
-			} else {
+
+			// For a compressed version of query plan, check if it is necessary to add the l2g operator over a request based on the form of the graph pattern.
+			// The current implementation assumes that only concepts and roles are being considered in vocabulary mapping.
+			boolean rewriteNeeded = false;
+			if ( compressVocabularyRewriting ) {
+				final Set<TriplePattern> tps = LogicalOpUtils.getTriplePatternsOfReq(requestOp);
+				if ( !tps.isEmpty() ) {
+					for ( final TriplePattern tp : tps ) {

```

Comment:
It would be better to take this whole code block that is checking whether `rewriteNeeded` should be true and move it to a separate function which you can then simply call from here.

Reply:
Additionally, this function should be called only after making sure that there is a vocabulary mapping. (currently, your code is always checking whether `rewriteNeeded` should be true, even in cases in which the fed.member of the request op. doesn't have a vocab.mapping).